parquet/src/main/java/org/apache/iceberg/parquet/ParquetReadSupport.java

```diff
@@ -55,9 +58,16 @@ public ReadContext init(Configuration configuration, Map<String, String> keyValu
     // matching to the file's columns by full path, so this must select columns by using the path
     // in the file's schema.
 
-    MessageType projection = ParquetSchemaUtil.hasIds(fileSchema) ?
-        ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
-        ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
+    MessageType projection;
+    if (ParquetSchemaUtil.hasIds(fileSchema)) {
+      projection = ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema);
+    } else {
+      if (nameMapping != null) {
+        projection = ParquetSchemaUtil.pruneColumnsByName(fileSchema, expectedSchema, nameMapping);
+      } else {
+        projection = ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
```

Comment:
Let's simplify this logic by adding `nameMapping` to `pruneColumnsFallback`. Then inside that method, we can decide the fallback strategy based on whether nameMapping is null.